lib/collection/src/collection/clean.rs

```diff
@@ -0,0 +1,318 @@
+    ) -> CollectionResult<UpdateStatus> {
+        let mut tasks = self.tasks.upgradable_read();
+
+        // Await existing task if not cancelled or failed
+        if let Some(task) = tasks
+            .get(&shard_id)
+            .filter(|task| !task.is_cancelled_or_failed())
+        {
+            let receiver = task.status.clone();
+            drop(tasks);
+            return Self::await_task(receiver, wait, timeout).await;
+        }
+
+        // Create and await new task
+        let receiver = tasks.with_upgraded(|tasks| {
+            let task = ShardCleanTask::new(shards_holder, shard_id);
+            let receiver = task.status.clone();
+            tasks.insert(shard_id, task);
+            receiver
+        });
+        drop(tasks);
+
+        Self::await_task(receiver, wait, timeout).await
+    }
+
+    /// Await for an ongoing task to finish by its status receiver
+    async fn await_task(
+        mut receiver: Receiver<ShardCleanStatus>,
+        wait: bool,
+        timeout: Option<Duration>,
+    ) -> CollectionResult<UpdateStatus> {
+        let start = Instant::now();
+
+        loop {
+            match receiver.borrow_and_update().deref() {
+                ShardCleanStatus::Started => {}
+                ShardCleanStatus::Done => return Ok(UpdateStatus::Completed),
+                ShardCleanStatus::Failed(err) => {
+                    return Err(CollectionError::service_error(format!(
+                        "failed to clean shard points: {err}",
+                    )))
+                }
+                ShardCleanStatus::Cancelled => {
+                    return Err(CollectionError::service_error(
+                        "failed to clean shard points due to cancellation, please try again",
+                    ))
+                }
+            }
+
+            if !wait {
+                return Ok(UpdateStatus::Acknowledged);
+            }
+
+            let result = if let Some(timeout) = timeout {
+                tokio::time::timeout(timeout - start.elapsed(), receiver.changed()).await
+            } else {
+                Ok(receiver.changed().await)
+            };
+            match result {
+                // Status updated, loop again to check it another time
+                Ok(Ok(_)) => continue,
+                // Channel dropped, return error
+                Ok(Err(_)) => {
+                    return Err(CollectionError::service_error(
+                        "Failed to clean shard points, notification channel dropped",
+                    ))
+                }
+                // Timeout elapsed, acknowledge so the client can probe again later
+                Err(_) => return Ok(UpdateStatus::Acknowledged),
+            }
+        }
+    }
+
+    /// Cancel cleaning of a shard and mark it as dirty
+    pub(super) fn invalidate(&self, shard_id: ShardId) {
+        let removed = self.tasks.write().remove(&shard_id);
+        if let Some(task) = removed {
+            task.abort();
```

Comment:
Here we are not `awaiting` the aborted task.
Is this intended?

Reply:
I initially thought - we don't really care.

But on second thought it may be better to await, thanks!

If we change the hash rings (and invalidate here), we don't want it to delete the wrong points based on the old hash ring. Awaiting here would resolve that.

That does mean it might await on a non-async lock, so I might replace it with a tokio RwLock instead.

I'll think about it for a bit and implement a fix.